web/pandas/pdeps/0006-ban-upcasting.md

```diff
@@ -0,0 +1,149 @@
+of operating inplace to begin with, such as:
+- ``ser.diff()``;
+- ``pd.concat([ser, other])``;
+- ``ser.mean()``.
+
+These would keep being allowed to change Series' dtypes.
+
+## Detailed description
+
+Concretely, the suggestion is:
+- if a ``Series`` is of a given dtype, then a ``setitem``-like operation should not change its dtype;
+- if a ``setitem``-like operation would previously have changed a ``Series``' dtype, it would now raise.
+
+For a start, this would involve:
+
+1. changing ``Block.setitem`` such that it doesn't have an ``except`` block in
+
+  ```python
+  value = extract_array(value, extract_numpy=True)
+  try:
+      casted = np_can_hold_element(values.dtype, value)
+  except LossySetitemError:
+      # current dtype cannot store value, coerce to common dtype
+      nb = self.coerce_to_target_dtype(value)
+      return nb.setitem(indexer, value)
+  else:
+  ```
+
+2. making a similar change in ``Block.where``, ``EABlock.setitem``, ``EABlock.where``, and probably more places.
+
+The above would already require several hundreds of tests to be adjusted.
+
+### Ban upcasting altogether, or just upcasting to ``object``?
+
+The trickiest part of this proposal concerns what to do when setting a float in an integer column:
+
+```python
+In [1]: ser = pd.Series([1, 2, 3])
+
+In [2]: ser[0] = 1.5
+```
+
+This isn't necessarily a sign of a bug, because the user might just be thinking of their ``Series`` as being
+numeric (without much regard for ``int`` vs ``float``) - ``'int64'`` is just what pandas happened to infer.
+
+Possibly options could be:
+1. just raise, forcing users to be explicit;
+2. convert the float value to ``int`` before setting it;
+3. limit "banning upcasting" to when the upcasted dtype is ``object``.
+
+Let us compare with what other libraries do:
+- ``numpy``: option 2
+- ``cudf``: option 2
+- ``polars``: option 2
+- ``R data.frame``: just upcasts (like pandas does now for non-nullable dtypes);
+- ``pandas`` (nullable dtypes): option 1
+- ``datatable``: option 1
+- ``DataFrames.jl``: option 1
+
+Option ``2`` would be a breaking behaviour change in pandas. Further,
+if the objective of this PDEP is to prevent bugs, then this is also not desirable:
+someone might set ``1.5`` and later be surprised to learn that they actually set ``1``.
+
+There are several downsides to option ``3``:
+- it would be inconsistent with the nullable dtypes' behaviour;
+- it would also add complexity to the codebase and to tests;
+- it would be hard to teach, as instead of being able to teach a simple rule,
+  there would be a rule with exceptions;
+- there would be a risk of loss of precision;
+- it opens the door to other exceptions, such as not upcasting to ``'int16'``
+  when trying to set an element of a ``'int8'`` ``Series`` to ``128``.
+
+Option ``1`` is the maximally safe one in terms of protecting users from bugs, being
+consistent with the current behaviour of nullable dtypes, and in being simple to teach.
```

Comment:
+1 on option 1.

Reply:
+1 for option 1 as well

Reply:
Option 1 has some downsides that aren't listed:
  1. How do we handle upcasting across different data sizes; ex: assigning a float32 "integral" value to an int8 - do we upcast to int8?
  2. Relatedly how do we handle "integral" float values that exceed the storage range of an integer?
  3. Extra performance penalty of checking if a float can be represented as an integral

None of these are blockers for this PDEP but something to still think about

Reply:
Option 1 could cause confusion for people who don't understand how floating point numbers work.  For example:
```python
>>> s = pd.Series([1,2,3])
>>> s[1] = 1.0000000000000001
>>> s
0    1
1    1
2    3
dtype: int64
>>> s[1] = 1.0000000000000002
>>> s
0    1.0
1    1.0
2    3.0
dtype: float64
```
In this proposal, setting to `1.0000000000000001` would be accepted, because this floating point number is rounded by python to 1.0), while `1.0000000000000002` would raise, because it is seen as different than 1.0.

It may be worthwhile to introduce a tolerance parameter (by default is "machine epsilon", i.e., `np.finfo(np.float64).eps`), that must be positive and less than 1.0 and greater than or equal to machine epsilon, where we say that anything that is within the tolerance is considered an integer.  So if the tolerance was `1.0e-6`, then in both of the above cases, we would treat the value as integral, and also treat `1.00000001` as integer as well.

NOTE:  This is a BIG Deal in my area of mathematical optimization, where we have integrality tolerances in solvers to decide if an answer to an optimization problem is integer-valued.

Reply:
> How do we handle upcasting across different data sizes; ex: assigning a float32 "integral" value to an int8 - do we upcast to int8?

It would raise (because the current behaviour would be to upcast:
```python
In [31]: ser
Out[31]:
0    1
1    2
2    3
dtype: int8

In [32]: ser[2] = 100000000.0

In [33]: ser
Out[33]:
0            1
1            2
2    100000000
dtype: int32
```

> Extra performance penalty of checking if a float can be represented as an integral

Sure, but that wouldn't be a regression on the status-quo, because that's what pandas already does:
```python
In [35]: ser
Out[35]:
0    1
1    2
2    3
dtype: int64

In [36]: ser[2] = 4.0

In [37]: ser
Out[37]:
0    1
1    2
2    4
dtype: int64
```

Reply:
Yep understood on status quo - thanks for responding

Reply:
> Option 1 could cause confusion for people who don't understand how floating point numbers work. 

@Dr-Irv I think the same is true for the other options as well? Because with option 3 (current behaviour) you also get a different result (different dtype) because of this small difference, and with option 2 you loose any decimal part, which kind of "avoids" it, but you might still care about losing this part or not, depending on how close the floating point was to the actual integer. 
I suppose only when completely disallowing to set a float value into an integer column (even for "round" float values), there is no confusion. But even then, if you leave it to the user to convert their float into an int in the way it makes most sense for them, they probably still need to understand those floating point details to use the "correct" method.

So I am not sure there is, in general, a way around people having to understand how floating point numbers work, if they run into cases where small epsilon-size differences matter (ànd if they then also want to interpret those numbers as integers).

> It may be worthwhile to introduce a tolerance parameter

For actual setitem indexing operations (eg `df.loc[idx, col] = val`), there is not an easy way to specify a parameter (except using a global option). 
If a user wants to control a custom tolerance, they can still preprocess values themselves before setting then into the pandas Series? (i.e. `df.loc[idx, col] = maybe_convert_to_int(val, tolerance=..)`)

Reply:
> If a user wants to control a custom tolerance, they can still preprocess values themselves before setting then into the pandas Series? (i.e. `df.loc[idx, col] = maybe_convert_to_int(val, tolerance=..)`)

Maybe we should be including a function like `maybe_convert_to_int` as part of this proposal?  (I'm not sure).  At least from a documentation standpoint, we will have to explain what is considered an int versus a float.  Unfortunately, most people don't understand this issue with tolerances.

Reply:
I've added this to the faq

Regarding `maybe_convert_to_int` - could you give an example of how you currently handle it?

I'd be inclined to leave it out though, it sounds like it something that would be quite domain-specific

Reply:
> Regarding `maybe_convert_to_int` - could you give an example of how you currently handle it?

Not sure if this will answer your question, but we might do things like this:
```python
TOLERANCE = 1.0e-8
rounded = s.mask((s-s.round()).abs() < TOLERANCE, s.round())
```

My point here is that if we are telling people that "If the value is a `float` that looks like an `int`, and the `Series` is an `int` dtype, pandas will not raise", then we have to tell people what "looks like an `int`" means.  If you're doing some computation, then maybe a value like 1.0+1.0e-13 "looks like" an `int`.  So providing a functionality like `maybe_convert_to_int()` as suggested by @jorisvandenbossche would create awareness of the issue and let people have control over what "looks like an `int`"

Reply:
ok, I've added an example of such a function